File: test/src/main/java/com/shipping/service/TestService.java
package com.shipping.service;

import com.shipping.domain.sys.User;
import com.shipping.repository.sys.UserRepository;
import org.spin.core.throwable.SimplifiedException;
import org.spin.web.annotation.RestfulMethod;
import org.spin.web.annotation.RestfulService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Required;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.multipart.MultipartFile;

import java.beans.Transient;
import java.util.List;

/**
 * <p>Created by xuweinan on 2017/9/17.</p>
 *
 * @author xuweinan
 */
@RestfulService("Test")
public class TestService {

    @Autowired
    private UserRepository userDao;

    @RestfulMethod(auth = false, authRouter = "a")
    public int aaa(String[] pp, List a, MultipartFile[] b) {
//        throw new SimplifiedException("aaa");
        return pp.length;
    }

    @Transactional
    @RestfulMethod(auth = false, value = "b")
    public void testTransaction() {
        User user = new User();
        user.setRealName("二傻子");
        user.setUserName("admin");
        user.setPassword("123");
        user.setMobile("13111111111");
        user.setEmail("none@qq.com");
        System.out.println(userDao.getCurrentDataSourceName());
        userDao.save(user);
        userDao.switchDataSource("db2");
        System.out.println(userDao.getCurrentDataSourceName());
        user = user.getDTO(1);
        user.setId(null);
        userDao.save(user);
        userDao.switchDataSource("db1");
//        throw new SimplifiedException("aa");
    }
}

File: test/src/main/kotlin/com/shipping/service/TestService.kt
package com.shipping.service

import com.shipping.domain.sys.User
import mu.KLogging
import org.spin.data.core.DataSourceContext
import org.spin.data.extend.RepositoryContext
import org.spin.web.annotation.RestfulMethod
import org.spin.web.annotation.RestfulService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.transaction.annotation.Transactional
import org.springframework.web.multipart.MultipartFile

/**
 *
 * Created by xuweinan on 2017/9/17.
 *
 * @author xuweinan
 */
@RestfulService("Test")
class TestService {
    companion object : KLogging()

    @Autowired
    private lateinit var repoCtx: RepositoryContext

    @RestfulMethod(auth = false, authRouter = "a")
    fun aaa(pp: Array<String>, a: List<*>?, b: Array<MultipartFile>?): Int {
        //        throw new SimplifiedException("aaa");
        println(a?.size)
        println(b?.size)
        logger.debug { "aaa" }
        return pp.size
    }

    @Transactional
    @RestfulMethod(auth = false, value = "b")
    fun testTransaction() {
        var user = User().apply {
            realName = "二傻子"
            userName = "admin"
            password = "123"
            mobile = "13111111111"
            email = "none@qq.com"
        }
        println(DataSourceContext.getCurrentDataSourceName())
        repoCtx.save(user)
        DataSourceContext.switchDataSource("db2")
        println(DataSourceContext.getCurrentDataSourceName())
        user = user.getDTO(1)
        user.id = null
        repoCtx.save(user)
        DataSourceContext.usePrimaryDataSource()
        //        throw new SimplifiedException("aa");
    }

    @Transactional
    @RestfulMethod(auth = false, value = "c")
    fun testTransaction2() {
        var user = User().apply {
            realName = "二傻子"
            userName = "admin"
            password = "123"
            mobile = "13111111111"
            email = "none@qq.com"
        }
        println(DataSourceContext.getCurrentDataSourceName())
        print(DataSourceContext.getCurrentSchema())
        repoCtx.save(user)
        DataSourceContext.switchSchema("db2")
        print(DataSourceContext.getCurrentSchema())
        user = user.getDTO(1)
        user.id = null
        repoCtx.save(user)
        DataSourceContext.restoreSchema()
        //        throw new SimplifiedException("aa");
    }

    @Transactional
    @RestfulMethod(auth = false, value = "d")
    fun testSTransaction() {
        val user = User().apply {
            realName = "xuweinan"
            userName = "admin"
            password = "123"
            mobile = "13111111111"
            email = "none@qq.com"
        }
        println(DataSourceContext.getCurrentDataSourceName())
        repoCtx.save(user)
        //        throw new SimplifiedException("aa");
        println(DataSourceContext.getCurrentSchema())
        DataSourceContext.switchSchema("dev_db2")
        println(DataSourceContext.getCurrentSchema())
        val u2 = User().apply {
            realName = "mathcat"
            userName = "admin"
            password = "123"
            mobile = "13111112222"
            email = "none@qq.com"
        }
        u2.id = null
        repoCtx.save(u2)
    }
}
